Add drag and drop to move items in FileTreeView

boopeshmahendran opened the pull request:
This PR adds the feature to drag and drop files and folders to another folder in the filetree to move them. We can drop items outside of all the folders to move items to the root folder.

I have tried to make the rename functions generic by making it accept new path instead of new name and  **reuse the rename workflow** for the move item as there is a lot common between the two things.

**Todo:-**
- [x] Add more tests
- [x] Refine the UI and fix some UI issues

@swmitra @nethip @petetnt @ficristo  @navch @shubhsnov  Please review

boopeshmahendran committed:
Add drag and drop to move items in FileTreeView

Todo:

* Handle move errors.

* Add support for moving to root directory.

boopeshmahendran commented on src/project/ProjectManager.js:
```diff
@@ -369,6 +377,34 @@ define(function (require, exports, module) {
+        var self = this;
+
+        // Remove selected marker
+        self.setSelected(null);
```
As of now, I just remove the selected flag if any of the items are moved. Need feedback on the correct way to handle it.

boopeshmahendran committed:
Add support for moving items to root directory

petetnt commented:
Really good stuff @boopeshmahendran!

Some preliminary things:

1. I think this falls under "Handle move errors", but trying to drop folder on itself brings up this:

   ![image](https://user-images.githubusercontent.com/7641760/28082722-ed65074e-667b-11e7-9aeb-5e420377a09f.png)
    I think it would make sense just to ignore that without showing any user errors.

2. When dragging and dropping files, it would make sense to uncollapse folders that are closed while dragging so that you can access deeply nested subfolders by just dragging. As in drag a file over a closed folder -> the folder uncollapses.

3. Not sure if you should be able to drop folders in top of the files too to move them. It's how Right click -> new folder works so it might make sense

petetnt commented on src/project/ProjectManager.js:
```diff
@@ -369,6 +377,34 @@ define(function (require, exports, module) {
+     * See `ProjectModel.moveItem`
+     */
+    ActionCreator.prototype.moveItem = function(oldPath, newDirectory) {
+        var self = this;
```
No need to set `this` to variable here (for now at least)

petetnt commented on src/project/ProjectModel.js:
```diff
@@ -1240,6 +1254,50 @@ define(function (require, exports, module) {
+        // Reusing the _renameItem for moving item
+        this._renameItem(oldPath, fullNewPath, fileName, !_pathIsFile(fullNewPath)).then(function () {
+            var newDirectoryRelative = self.makeProjectRelativeIfPossible(newDirectory),
+                needsLoading    = !self._viewModel.isPathLoaded(newDirectoryRelative);
```
Nit: some extra spaces after `needsLoading`

petetnt replied:
Maybe save the reference to `var viewModel = self._viewModel` to make the calls a bit less terse.

petetnt commented on src/project/ProjectModel.js:
```diff
@@ -1240,6 +1254,50 @@ define(function (require, exports, module) {
+            if (needsLoading) {
+                self._getDirectoryContents(newDirectory).then(function(contents) {
+                    self._viewModel.setDirectoryContents(newDirectoryRelative, contents);
+                    self._viewModel.moveItem(self.makeProjectRelativeIfPossible(oldPath), self.makeProjectRelativeIfPossible(fullNewPath));
```
Maybe make these variables too `self.makeProjectRelativeIfPossible(oldPath)` `self.makeProjectRelativeIfPossible(fullNewPath)` as the long lines are pretty hard to distinquish.

petetnt reviewed (commented):
Some really small comments, really nice job 👍

boopeshmahendran commented:
@petetnt Thanks for reviewing. I'll address these comments.

boopeshmahendran committed:
Check item dropped onto itself or parent directory

boopeshmahendran committed:
Add support for moving items by dropping on files

boopeshmahendran committed:
Create dragItem action

boopeshmahendran committed:
Close directory on drag

boopeshmahendran committed:
Open directory on Drop

boopeshmahendran committed:
Open directory on drag over

boopeshmahendran committed:
Address review comments

boopeshmahendran committed:
Add tests for moveItem in FileTreeViewModel

boopeshmahendran commented:
@petetnt I have addressed your comments. Please review.

boopeshmahendran committed:
Fix style issues on drag

boopeshmahendran committed:
Make styles fast

boopeshmahendran committed:
Change fileindex to update the moved entry

boopeshmahendran committed:
Fix lint mistakes

boopeshmahendran committed:
Set drag image as item name

boopeshmahendran committed:
Check if directory is open before opening directory on drop

boopeshmahendran committed:
Refactor code

boopeshmahendran committed:
Check if item is dropped onto itself or parent directory

boopeshmahendran committed:
Move selected flag when item is moved

boopeshmahendran committed:
Merge branch 'master' into dragAndDrop

boopeshmahendran committed:
Merge branch 'master' into dragAndDrop

boopeshmahendran committed:
Use filter instead of forEach

boopeshmahendran committed:
Change implementation to reuse rename workflow

boopeshmahendran committed:
Fix tests

boopeshmahendran committed:
Add docs and comments

boopeshmahendran committed:
Address review comments

boopeshmahendran commented on src/filesystem/FileIndex.js:
```diff
@@ -138,6 +142,30 @@ define(function (require, exports, module) {
+
+            if (oldDirectory && oldDirectory._contents) {
+                oldDirectory._contents = oldDirectory._contents.filter(function(entry) {
+                    if (entry.fullPath === oldPath || entry.fullPath === newPath) {
```
Yeah. Changed this.

petetnt reviewed (commented):
Codewise LGTM, didn't have the time to re-test the functionality yet

boopeshmahendran committed:
Add Error handling

boopeshmahendran committed:
Make directory open and adding new item independent

This makes adding the moved item to the new directory independent of
whether the directory is loaded or not by creating a notFullyLoaded
directory for the new directory.
